File: backend/src/main/java/com/apocalypse/thefall/controller/CharacterController.java
package com.apocalypse.thefall.controller;

import com.apocalypse.thefall.dto.CharacterCreationDto;
import com.apocalypse.thefall.dto.CharacterDto;
import com.apocalypse.thefall.dto.CharacterInventoryDto;
import com.apocalypse.thefall.dto.MoveRequestDto;
import com.apocalypse.thefall.mapper.CharacterMapper;
import com.apocalypse.thefall.service.CharacterService;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/characters")
@RequiredArgsConstructor
public class CharacterController {

    private final CharacterService characterService;
    private final CharacterMapper characterMapper;

    @PostMapping
    public CharacterDto createCharacter(@AuthenticationPrincipal Jwt jwt, @RequestBody CharacterCreationDto request) {
        return characterMapper.toDto(characterService.createCharacter(jwt.getSubject(), characterMapper.fromCreationDto(request)));
    }

    @GetMapping("/current")
    public CharacterDto getCurrentCharacter(@AuthenticationPrincipal Jwt jwt) {
        return characterMapper.toDto(characterService.getCharacterByUserId(jwt.getSubject()));
    }

    @PostMapping("/move")
    public CharacterInventoryDto moveCharacter(@AuthenticationPrincipal Jwt jwt, @RequestBody MoveRequestDto request) {
        return characterMapper.toCharacterInventoryDto(characterService.moveCharacter(jwt.getSubject(), request.x(), request.y()));
    }
}

File: backend/src/main/java/com/apocalypse/thefall/controller/character/CharacterController.kt
package com.apocalypse.thefall.controller.character

import com.apocalypse.thefall.dto.MoveRequestDto
import com.apocalypse.thefall.dto.character.CharacterCreationDto
import com.apocalypse.thefall.dto.character.CharacterDto
import com.apocalypse.thefall.mapper.character.toDto
import com.apocalypse.thefall.mapper.character.toEntity
import com.apocalypse.thefall.service.character.CharacterService
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.security.oauth2.jwt.Jwt
import org.springframework.web.bind.annotation.*


@RestController
@RequestMapping("/api/characters")
class CharacterController(
    private val characterService: CharacterService
) {

    @PostMapping
    fun createCharacter(
        @AuthenticationPrincipal jwt: Jwt,
        @RequestBody request: CharacterCreationDto
    ): CharacterDto =
        characterService.createCharacter(jwt.subject, request.toEntity()).toDto()

    @GetMapping("/current")
    fun getCurrentCharacter(@AuthenticationPrincipal jwt: Jwt): CharacterDto =
        characterService.getSimpleCharacterByUserId(jwt.subject).toDto()

    @PostMapping("/move")
    fun moveCharacter(
        @AuthenticationPrincipal jwt: Jwt,
        @RequestBody request: MoveRequestDto
    ): CharacterDto = characterService.moveCharacter(jwt.subject, request.x, request.y).toDto()
}